File: Modules/Common/Techs/SIE.Web.Tech/Routings/Commands/SaveRoutingCommand.cs
using SIE.Domain;
using SIE.Security;
using SIE.Tech.Routings;
using SIE.Tech.Routings.ViewModels;
using SIE.Web.Command;

namespace SIE.Web.Tech.Routings.Commands
{
    /// <summary>
    /// 保存工艺路线
    /// </summary>
    [JsCommand("SIE.Web.Tech.Routings.Commands.SaveRoutingCommand")]
    [AllowAnonymous]
    public class SaveRoutingCommand : ViewCommand
    {
        /// <summary>
        /// 保存工艺路线执行方法
        /// </summary>
        /// <param name="args">视图参数</param>
        /// <param name="scope">作用域</param>
        /// <returns>保存结果</returns>
        protected override object Excute(ViewArgs args, string scope)
        {
            var obj = args.Data.ToJsonObject<RoutingLayoutMsg>();
            if (obj.RoutingId > 0)
            {
                if (obj.RoutingVersionId > 0)
                {
                    var version = RF.Find<RoutingVersion>().GetById(obj.RoutingVersionId) as RoutingVersion;
                    if (version != null)
                    {
                        version.Layout.Layout = obj.Layout;
                        RF.Save(version.Layout);
                    }
                }
                else
                {
                    var version = RT.Service.Resolve<RoutingController>().CreateRoutingVersion(obj);
                    string name = (version.IsDefault == YesNo.Yes ? "*" : string.Empty) + version.Name + "(" + version.ReferenceQty + ")";
                    int.TryParse(version.Name.Substring(1), out int number);
                    //新增工艺路线版本返回新增的节点，让前端展开显示（前端判断非true则增加节点并展开显示）
                    return new
                    {
                        Id = version.Id,
                        text = name,
                        leaf = true,
                        nodetype = "VersionNode",
                        routingId = version.RoutingId,
                        layoutId = version.LayoutId,
                        state = version.State,
                        isDefault = version.IsDefault,
                        referenceTime = version.ReferenceQty,
                        versionName = version.Name,
                        versionNum = number
                    };
                }
            }
            //其他return true
            return true;
        }
    }
}

File: Modules/Common/Techs/SIE.Web.Tech/Routings/Commands/SaveRoutingCommand.js
SIE.defineCommand('SIE.Web.Tech.Routings.Commands.SaveRoutingCommand', {
    extend: 'SIE.cmd.Command',
    meta: { text: "保存", group: "edit", tooltip: "保存工艺路线".t(), iconCls: "icon-PlaylistCheck icon-blue" },
    execute: function (view, source) {
        var version = view.CurRoutingVersion;
        var routingId = version.data.routingId;
        var routingVersionId = version.data.Id;

        
        var layout = Ext.getCmp(view.routingDrawControlId).getXml();

        var iptdata = {
            RoutingId: routingId,
            RoutingVersionId: routingVersionId,
            Layout: layout
        };
        //判断是否复制粘贴保存       
        if (version.data.isCopy || version.data.isNew) {
            iptdata.RoutingId = version.data.targetRoutingId;
            iptdata.VersionName = version.data.versionName;
            iptdata.RoutingVersionId = 0;
        }
        //客制化界面命令注册签名（注册在context，没有后台请求）
        SIE.Signature.createCmdContext({ command: Ext.getClassName(this), commandName: "保存", Type: view.model });
        view.execute({
            command: Ext.getClassName(this),
            data: iptdata,
            success: function (res) {
                SIE.Msg.showInstantMessage('保存成功'.t());
                if (res.Result !== true) {
                    //复制粘贴保存后，展开节点                  
                    var routingVersion = res.Result;
                    var routingNode = version.parentNode;
                    if (routingVersion.versionNum)
                        routingNode.data.MaxVersionNum = routingVersion.versionNum;
                    routingNode.removeChild(version);
                    if (!routingNode) {
                        var routingControl = view.layout.routingControl;
                        routingNode = routingControl.getRootNode().childNodes.selectMany(function (p) { return p.childNodes; }).first(function (p) { return p.data.Id === routingVersion.routingId; });
                    }
                    var node = routingNode.appendChild(routingVersion);
                    var treePanel = routingNode.getOwnerTree();
                    treePanel.getSelectionModel().select(node);
                    var treeview = treePanel.getView();
                    if (routingNode && routingNode.parentNode) {
                        routingNode.expand();
                        routingNode.parentNode.expand();
                    }
                    treeview.focusRow(node);
                    treeview.fireEvent('itemdblclick', treeview, node, treeview.getNodeByRecord(node), treeview.indexOfRow(node));
                }
            }
        });
    }
});
